### ISSUE TITLE
Add integration with asyncio

### ISSUE THREAD

**Issue by fxstein** [datafolklabs/cement#296]
Would love to use cement for a framework project I am working on, but it would need to support asyncio beyond just placing the run_loop into a handler. Not sure if anybody has looked at what it would mean to make cement fully async. Would be a very powerful combination.

**Comment by derks** [comment]
Would be very interested in this, though have very little experience with asyncio.  Could you provide any further detail in what your looking for in regards to Cement?  Feel free to email me directly if you want to spitball thoughts "offline".  derks [at] datafolklabs [dot] com

**Comment by derks** [comment]
Per discussions we had "offline" the majority of Cement's non-async features are on foundation, and "startup".  Building an app with asyncio should not be completely limited by Cement, but is entirely a per-use case situation.  For that reason, I am closing this out until future issues arise that are more specific and have some actionable refactoring tied to them.

**Comment by fxstein** [comment]
I believe I finally got to the point where extra features for asyncio are required.

The example I have now encountered are in regards to application hooks in plugins. I have an event processing hook that can be defined per plugin. With the default cement behavior these are functions and not generators. As such you cannot yield from them and in return they are synchronous code.

That means that any delay caused by them will stall the entire application.

When defining a hook as a generator via

````
@asyncio.coroutine
def my_hook():
````
cement fails with an exception in cement/core/hook.py line 118:

````
yield res
````
as python cannot yield a generator result but rather has to yield from a generator:

````
yield from res
````
In order to support both functions and generator cement/core/hook.py could be modified to attempt a yield from in case the hook is a generator and fallback to yield for regular functions:

````
try:
    yield from res
except TypeError:
    yield res
````
With this minor modification I have been able to execute both function as well as generator hooks, allowing me to process out of sync or in a separate thread or perform 
````
asyncio.sleep(x)
````
inside the generator hook without blocking the rest of the application.

I will submit a pull request with this suggested change. Thanks!
